prov/shm/src/smr.h

```diff
@@ -201,8 +217,11 @@ struct smr_ep {
 	struct smr_queue	trecv_queue;
 	struct smr_unexp_fs	*unexp_fs;
 	struct smr_pend_fs	*pend_fs;
+	struct smr_sar_fs	*sar_fs;
 	struct smr_queue	unexp_msg_queue;
 	struct smr_queue	unexp_tagged_queue;
+	struct dlist_entry	sar_list;
+	enum smr_peer_state	state[SMR_MAX_PEERS];
```

Comment:
Ideally it would be. The problem is that the smr_ep doesn't currently hold any per peer information. Either it's in the av or it's shared in the EP memory region. I'm fine moving this to the shared memory region. The main reason I didn't is because it doesn't need to be shared by the peers and the memory is called "peer_addr" which this isn't really. But I'm fine either renaming or looking the other way.
But this is also the general source of my comment earlier that to properly implement this (and to fix the shm provider) we would need to refactor the entire addressing/peer exchange structure to have all the correct peer data (including cmd_cnt, sar_cnt) in the shared memory per-peer data. The reason we can't truly do it right now (which I realized while implementing this) is that the target peer is not always guaranteed to have the initiating peer inserted/mapped which means none of this peer_addr data is going to be valid and also means transfers that require a response entry (iov, mmap, sar, or DELIVERY_COMPLETE) will fail if the peer has not inserted the initiator into the AV yet.
I'm working on fixing this by making sure the peer maps the shared memory with a RTS/CTS type of protocol (sort of like how rxd exchanges the dg addresses) 
I know that was a lot of words... feel free to message me for alternate words.

Reply:
okee-dokee